draft/ZEP0010.md

```diff
@@ -0,0 +1,413 @@
+
+To provide for more flexible, immediate, and de-centralized use cases, we
+propose to add a generic extension point ```extensions`` on
+both arrays and groups into which extensions MAY be added.
+
+The ``extensions`` field if present MUST contain an array of extension
+definitions. The contained array MUST either have one or more extensions or the
+object MUST be omitted entirely.
+
+Further details on the specification changes can be found in
+<https://github.com/zarr-developers/zarr-specs/pull/344>.
+
+### Definition and naming
+
+Each extension object will follow the rules laid out in the "Zarr extensions"
+section of the v3 specification.
+
+### Processing
+
+In a manner similar to OpenGL’s extension mechanism, where implementations must
+explicitly advertise and support extensions they recognize, Zarr implementers
+are expected to inspect the extensions array and determine whether each listed
+extension is supported. If an extension includes ``"must_understand": true``
+(the default) and the implementation does not support it, the dataset must not
+be loaded and an appropriate error should be raised. For extensions with
+``"must_understand": false``, implementers may safely ignore unrecognized entries.
+
+To support a given extension, an implementation must either (1) check for known
+extension names and invoke appropriate logic according to the extension’s
+specification at the correct point in its processing pipeline (e.g., during
+metadata interpretation, data access, or layout resolution), or (2) delegate
+that logic via a callback or plugin mechanism that allows third-party code to
+handle the extension dynamically. This modular approach enables implementers to
+support a flexible and evolving set of extensions while maintaining core
+compatibility.
+
+### Examples
+
+The following examples represent a few realistic use cases of the top-level
+``extensions`` container. This ZEP is putting in place the mechanism so the
+community can experiment with such extensions before their standardization.
+
+#### Offset (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.offset",
+            "configuration": { "offset": [ 12, 24 ] }
+        }
+    ]
+}
+```
+
+The ``example.offset`` extension contains an array of the same order as the
+shape of the containing array specifying which element of the array should be
+considered as the origin, e.g., `[0, 0]`. This allows the reuse of subregions
+of an array without the need to rewrite the data.
+
+Note that in this example of the extension is ``must_understand=true`` meaning
+an implementation which does not support the ``example.offset`` extension
+should raise an error.
+
+#### Statistics (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.array-statistics",
+            "must_understand": false,
```

Comment:
Here we run into the issue that we may want to distinguish between reading and writing.  Reading the array without understanding the extension is fine, but writing to the array would invalidate the statistics.  Perhaps we should take the opportunity here to introduce `must_understand_for_reading` and `must_understand_for_writing`.  An extension marked plain `must_understand: false` would seem to be serving the same purpose as an attribute.

Reply:
Remembering your previous comment to this (...somewhere), I did consider it. I'm up for trying to include it now, but it will push the timeline, so there's a trade-off. As long as we're discussing it here, though, I was wondering if perhaps we don't make it an (extendable?!) object: `must_understand: {"read": true, "write" false}`.